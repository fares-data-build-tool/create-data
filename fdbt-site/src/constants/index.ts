export const OPERATOR_COOKIE = 'fdbt-operator';

export const FARE_TYPE_COOKIE = 'fdbt-fare-type';

export const PASSENGER_TYPE_COOKIE = 'fdbt-passenger-type';

export const SERVICE_COOKIE = 'fdbt-service';

export const JOURNEY_COOKIE = 'fdbt-journey';

export const GOVUK_LINK = 'https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/';

export const FARE_STAGES_COOKIE = 'fdbt-fare-stages';

export const STAGE_NAME_VALIDATION_COOKIE = 'fdbt-stage-names-validation';

export const FEEDBACK_LINK = 'mailto:fdbt@transportforthenorth.com?bcc=tfn-feedback@infinityworks.com';

export const STAGE_NAMES_COOKIE = 'fdbt-stage-names';

export const DAYS_VALID_COOKIE = 'fdbt-days-valid';

export const PRODUCT_DETAILS_COOKIE = 'fdbt-product-details';

export const PERIOD_TYPE_COOKIE = 'fdbt-period-type';

export const CSV_UPLOAD_COOKIE = 'fdbt-csv-upload';

export const CSV_ZONE_UPLOAD_COOKIE = 'fdbt-csv-zone-upload';

export const PERIOD_EXPIRY_COOKIE = 'fdbt-period-expiry';

export const SERVICE_LIST_COOKIE = 'fdbt-services';

export const INPUT_METHOD_COOKIE = 'fdbt-input-method';

export const NUMBER_OF_STAGES_COOKIE = 'fdbt-number-stages';

export const MATCHING_COOKIE = 'fdbt-matching';

export const MULTIPLE_SERVICE_COOKIE = 'fdbt-multiple-service';

export const NUMBER_OF_PRODUCTS_COOKIE = 'fdbt-number-of-products';

export const MULTIPLE_PRODUCT_COOKIE = 'fdbt-multiple-product';

export const ALL_COOKIES: string[] = [
    'fdbt-operator',
    'fdbt-fare-type',
    'fdbt-passenger-type',
    'fdbt-service',
    'fdbt-journey',
    'fdbt-fare-stages',
    'fdbt-stage-names-validation',
    'fdbt-stage-names',
    'fdbt-price-entry',
    'fdbt-days-valid',
    'fdbt-product-details',
    'fdbt-period-type',
    'fdbt-csv-upload',
    'fdbt-csv-zone-upload',
    'fdbt-number-of-products',
    'fdbt-period-expiry',
    'fdbt-services',
    'fdbt-input-method',
    'fdbt-matching',
    'fdbt-number-stages',
    'fdbt-multiple-service',
    'fdbt-number-of-products',
    'fdbt-multiple-product',
];

export const ALLOWED_CSV_FILE_TYPES = [
    'text/plain',
    'text/x-csv',
    'application/vnd.ms-excel',
    'application/csv',
    'application/x-csv',
    'application/octet-stream',
    'text/csv',
    'text/comma-separated-values',
    'text/x-comma-separated-values',
    'text/tab-separated-values',
];

export const STAGE = process.env.STAGE || 'dev';

export const RAW_USER_DATA_BUCKET_NAME = `fdbt-raw-user-data-${STAGE}`;
export const USER_DATA_BUCKET_NAME = `fdbt-user-data-${STAGE}`;
export const MATCHING_DATA_BUCKET_NAME = `fdbt-matching-data-${STAGE}`;
