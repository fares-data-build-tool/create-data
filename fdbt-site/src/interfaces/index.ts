export interface PeriodProductType {
    productName: string;
    productPrice: string;
    productNameError: boolean;
    productPriceError: boolean;
}
