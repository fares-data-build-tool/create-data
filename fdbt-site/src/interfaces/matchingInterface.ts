import { Stop } from '../data/auroradb';

export interface Price {
    price: string;
    fareZones: string[];
}

export interface MatchingFareZonesData {
    name: string;
    stops: Stop[];
    prices: Price[];
}

export interface MatchingFareZones {
    [key: string]: MatchingFareZonesData;
}
