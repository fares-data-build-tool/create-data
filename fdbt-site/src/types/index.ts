export interface ErrorInfo {
    errorMessage: string;
    id: string;
}
