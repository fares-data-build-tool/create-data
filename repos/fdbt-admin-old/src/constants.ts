export const ATTRIBUTE_MAP: { [key: string]: string } = {
    'custom:noc': 'NOC',
    'custom:schemeOperator': 'Scheme Name',
    'custom:schemeRegionCode': 'Scheme Region',
};

export const STATUS_MAP: { [key: string]: string } = {
    CONFIRMED: 'Registered',
    FORCE_CHANGE_PASSWORD: 'Awaiting Registration',
};

export const MAIN_USER_POOL_PREFIX = 'fdbt-user-pool';
export const NETEX_DATA_BUCKET_PREFIX = 'fdbt-netex-data';
export const PRODUCTS_DATA_BUCKET_PREFIX = 'fdbt-products-data';

export const AWS_REGION = 'eu-west-2';
