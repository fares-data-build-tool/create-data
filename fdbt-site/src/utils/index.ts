import Cookies from 'cookies';
import { NextPageContext } from 'next';
import { IncomingMessage } from 'http';
import axios from 'axios';
import { parseCookies, destroyCookie } from 'nookies';
import { decode } from 'jsonwebtoken';
import { OPERATOR_COOKIE, ID_TOKEN_COOKIE, REFRESH_TOKEN_COOKIE } from '../constants/index';
import { Stop } from '../data/auroradb';
import { ErrorInfo, CognitoIdToken } from '../interfaces';

export const setCookieOnServerSide = (ctx: NextPageContext, cookieName: string, cookieValue: string): void => {
    if (ctx.req && ctx.res) {
        const cookies = new Cookies(ctx.req, ctx.res);
        const host = ctx?.req?.headers?.host;
        const domain = host ? host.split(':')[0] : '';

        cookies.set(cookieName, cookieValue, { domain, path: '/' });
    }
};

export const deleteCookieOnServerSide = (ctx: NextPageContext, cookieName: string): void => {
    if (ctx.req && ctx.res) {
        const cookies = new Cookies(ctx.req, ctx.res);
        const host = ctx?.req?.headers?.host;
        const domain = host ? host.split(':')[0] : '';

        cookies.set(cookieName, '', { overwrite: true, maxAge: 0, domain, path: '/' });
    }
};

export const deleteAllCookiesOnServerSide = (ctx: NextPageContext): void => {
    const cookies = parseCookies(ctx);
    const cookieWhitelist = [OPERATOR_COOKIE, ID_TOKEN_COOKIE, REFRESH_TOKEN_COOKIE];

    Object.keys(cookies).forEach(cookie => {
        if (!cookieWhitelist.includes(cookie)) {
            destroyCookie(ctx, cookie);
        }
    });
};

export const getHost = (req: IncomingMessage | undefined): string => {
    if (!req) {
        return '';
    }
    const host = req?.headers?.host;

    if (host) {
        if (host && host.startsWith('localhost')) {
            return `http://${host}`;
        }
        return `https://${host}`;
    }

    return '';
};

export const isSessionValid = async (url: string, req: IncomingMessage | undefined): Promise<boolean> => {
    try {
        const response = await axios.get(url, {
            withCredentials: true,
            headers: {
                Cookie: req?.headers.cookie,
            },
        });

        return response.data.Valid;
    } catch (error) {
        return false;
    }
};

export const getUuidFromCookies = (ctx: NextPageContext): string | null => {
    const cookies = parseCookies(ctx);
    const operatorCookie = cookies[OPERATOR_COOKIE];
    if (!operatorCookie) {
        return null;
    }
    const operatorInfo = JSON.parse(operatorCookie);
    return operatorInfo.uuid;
};

export const getJourneyPatternFromCookies = (ctx: NextPageContext): string | null => {
    const cookies = parseCookies(ctx);
    const operatorCookie = cookies[OPERATOR_COOKIE];
    if (!operatorCookie) {
        return null;
    }
    const operatorInfo = JSON.parse(operatorCookie);
    return operatorInfo.journeyPattern;
};

export const formatStopName = (stop: Stop): string =>
    `${stop.localityName ? `${stop.localityName}, ` : ''}${stop.indicator ?? ''} ${stop.stopName ?? ''}${
        stop.street ? ` (on ${stop.street})` : ''
    }`;

export const buildTitle = (errors: ErrorInfo[], title: string): string => {
    if (errors.length > 0) {
        return `Error: ${title}`;
    }

    return title;
};

export const getAttributeFromIdToken = <T extends keyof CognitoIdToken>(
    ctx: NextPageContext,
    attribute: T,
): CognitoIdToken[T] | null => {
    const cookies = parseCookies(ctx);
    const idToken = cookies[ID_TOKEN_COOKIE];

    if (!idToken) {
        return null;
    }

    const decodedIdToken = decode(idToken) as CognitoIdToken;

    return decodedIdToken[attribute] ?? null;
};

export const getNocFromIdToken = (ctx: NextPageContext): string | null => getAttributeFromIdToken(ctx, 'custom:noc');
