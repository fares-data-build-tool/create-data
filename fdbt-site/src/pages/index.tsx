import React, { ReactElement } from 'react';
import TwoThirdsLayout from '../layout/Layout';
import { CustomAppProps, NextPageContextWithSession } from '../interfaces';
import { checkIfMultipleOperators } from '../utils';

const title = 'Create Fares Data';
const description = 'Create Fares Data is a service that allows you to generate data in NeTEx format';

interface StartProps {
    multipleOperators: boolean;
}

const Start = ({ multipleOperators }: StartProps & CustomAppProps): ReactElement => (
    <TwoThirdsLayout title={title} description={description}>
        <h1 className="govuk-heading-xl">Create Fares Data</h1>

        <p className="govuk-body">
            This service is for creating or accessing NeTEx data for public transport services, excluding rail, in
            England. The service can be used by:
        </p>
        <ul className="govuk-list govuk-list--bullet">
            <li>Bus operators running commercial services in England</li>
            <li>Local transport authorities acting on behalf of bus operators</li>
            <li>Local transport authorities that operate their own services</li>
        </ul>

        <p className="govuk-body">Use this service to:</p>
        <ul className="govuk-list govuk-list--bullet">
            <li>Generate fares data in NeTEx format for a new or existing service</li>
            <li>Download your own fares data</li>
        </ul>
        <p className="govuk-body">
            All service data is taken directly from the Traveline National Data Set and the NaPTAN database. To avoid
            data discrepancies, ensure that all service data is correct within these datasets before using the service.
        </p>

        <a
            href={multipleOperators ? '/multipleOperators' : '/fareType'}
            role="button"
            draggable="false"
            className="govuk-button govuk-button--start"
            data-module="govuk-button"
            id="start-now-button"
        >
            Start now
            <svg
                className="govuk-button__start-icon"
                xmlns="http://www.w3.org/2000/svg"
                width="17.5"
                height="19"
                viewBox="0 0 33 40"
                role="presentation"
                focusable="false"
            >
                <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z" />
            </svg>
        </a>
    </TwoThirdsLayout>
);

export const getServerSideProps = (ctx: NextPageContextWithSession): { props: StartProps } => ({
    props: { multipleOperators: checkIfMultipleOperators(ctx) },
});

export default Start;
