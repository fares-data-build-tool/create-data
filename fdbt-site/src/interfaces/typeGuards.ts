import {
    SingleTicketProps,
    ReturnTicketProps,
    PeriodTicketProps,
    FlatFareTicketProps,
} from '../pages/ticketConfirmation';
import { SelectSalesOfferPackageWithError } from '../pages/api/selectSalesOfferPackage';
import {
    PeriodTicket,
    PointToPointTicket,
    PeriodMultipleServicesTicket,
    FlatFareTicket,
    GeoZoneTicket,
    InputMethodInfo,
    ErrorInfo,
    Journey,
    JourneyWithErrors,
    TicketRepresentationAttribute,
    TicketRepresentationAttributeWithErrors,
    SalesOfferPackage,
    ProductWithSalesOfferPackages,
    TicketPeriod,
    MultiOperatorInfo,
    MultiOperatorInfoWithErrors,
} from './index';

import { FareType, FareTypeWithErrors } from '../pages/api/fareType';
import { PassengerType, PassengerTypeWithErrors } from '../pages/api/passengerType';
import { Service, ServiceWithErrors } from '../pages/api/service';
import { FareStagesAttribute, FareStagesAttributeWithErrors } from '../pages/api/chooseStages';
import { InputCheck } from '../pages/stageNames';
import { TicketPeriodWithErrors } from '../pages/api/productDateInformation';
import { MultipleOperatorsAttribute, MultipleOperatorsAttributeWithErrors } from '../pages/api/searchOperators';

export const isNotEmpty = <T>(value: T | null | undefined): value is T => value !== null && value !== undefined;

export const isPeriodTicket = (ticket: PeriodTicket | PointToPointTicket): ticket is PeriodTicket =>
    (ticket as PeriodTicket).products?.[0]?.productName !== undefined;

export const isMultipleServicesTicket = (
    ticket: PeriodTicket | PointToPointTicket,
): ticket is PeriodMultipleServicesTicket | FlatFareTicket =>
    (ticket as PeriodMultipleServicesTicket).selectedServices !== undefined;

export const isPointToPointTicket = (ticket: PeriodTicket | PointToPointTicket): ticket is PointToPointTicket =>
    (ticket as PointToPointTicket).lineName !== undefined;

export const isGeoZoneTicket = (ticket: PeriodTicket | PointToPointTicket): ticket is GeoZoneTicket =>
    (ticket as GeoZoneTicket).zoneName !== undefined;

export const isFareTypeAttributeWithErrors = (
    fareTypeAttribute: FareType | FareTypeWithErrors,
): fareTypeAttribute is FareTypeWithErrors => (fareTypeAttribute as FareTypeWithErrors).errors !== undefined;

export const isFareType = (fareType: FareType | FareTypeWithErrors | undefined): fareType is FareType => {
    return fareType !== undefined && (fareType as FareType).fareType !== undefined;
};

export const inputMethodErrorsExist = (
    inputMethodAttribute: InputMethodInfo | ErrorInfo | undefined,
): inputMethodAttribute is ErrorInfo => (inputMethodAttribute as ErrorInfo)?.errorMessage !== undefined;

export const isPassengerTypeAttributeWithErrors = (
    fareTypeAttribute: PassengerType | PassengerTypeWithErrors | undefined,
): fareTypeAttribute is PassengerTypeWithErrors => (fareTypeAttribute as PassengerTypeWithErrors).errors !== undefined;

export const isPassengerType = (
    passengerType: PassengerType | PassengerTypeWithErrors | undefined,
): passengerType is PassengerType => {
    return passengerType !== undefined && (passengerType as PassengerType).passengerType !== undefined;
};

export const isServiceAttributeWithErrors = (
    serviceAttribute: Service | ServiceWithErrors,
): serviceAttribute is ServiceWithErrors => (serviceAttribute as ServiceWithErrors).errors !== undefined;

export const isService = (service: Service | ServiceWithErrors | undefined): service is Service => {
    return service !== undefined && (service as Service).service !== undefined;
};

export const isJourneyAttributeWithErrors = (
    journeyAttribute: Journey | JourneyWithErrors,
): journeyAttribute is JourneyWithErrors => (journeyAttribute as JourneyWithErrors).errors !== undefined;

export const isJourney = (journey: Journey | JourneyWithErrors | undefined): journey is Journey => {
    return (
        journey !== undefined &&
        ((journey as Journey).directionJourneyPattern !== undefined ||
            (journey as Journey).inboundJourney !== undefined ||
            (journey as Journey).outboundJourney !== undefined ||
            (journey as Journey).errors !== undefined)
    );
};

export const isTicketRepresentationWithErrors = (
    ticketType: TicketRepresentationAttribute | TicketRepresentationAttributeWithErrors | undefined,
): ticketType is TicketRepresentationAttributeWithErrors =>
    ticketType !== undefined && (ticketType as TicketRepresentationAttributeWithErrors).errors !== undefined;

export const isTicketRepresentation = (
    ticketType: TicketRepresentationAttribute | TicketRepresentationAttributeWithErrors | undefined,
): ticketType is TicketRepresentationAttribute =>
    ticketType !== undefined && (ticketType as TicketRepresentationAttribute).name !== undefined;

export const isInputCheck = (stageNamesInfo: string[] | InputCheck[] | undefined): stageNamesInfo is InputCheck[] =>
    stageNamesInfo !== undefined && (stageNamesInfo[0] as InputCheck).error !== undefined;

export const isFareStageWithErrors = (
    fareStages: FareStagesAttribute | FareStagesAttributeWithErrors | undefined,
): fareStages is FareStagesAttributeWithErrors =>
    fareStages !== undefined && (fareStages as FareStagesAttributeWithErrors).errors !== undefined;

export const isFareStage = (
    fareStages: FareStagesAttribute | FareStagesAttributeWithErrors | undefined,
): fareStages is FareStagesAttribute =>
    fareStages !== undefined && (fareStages as FareStagesAttribute).fareStages !== undefined;

export const isSalesOfferPackageWithErrors = (
    salesOfferPackageInfo:
        | SelectSalesOfferPackageWithError
        | SalesOfferPackage[]
        | ProductWithSalesOfferPackages[]
        | undefined,
): salesOfferPackageInfo is SelectSalesOfferPackageWithError =>
    salesOfferPackageInfo !== undefined &&
    (salesOfferPackageInfo as SelectSalesOfferPackageWithError).errors !== undefined;

export const isProductWithSalesOfferPackages = (
    salesOfferPackageInfo:
        | SelectSalesOfferPackageWithError
        | SalesOfferPackage[]
        | ProductWithSalesOfferPackages[]
        | undefined,
): salesOfferPackageInfo is ProductWithSalesOfferPackages[] =>
    salesOfferPackageInfo !== undefined &&
    (salesOfferPackageInfo as ProductWithSalesOfferPackages[])[0].productName !== undefined;

export const isSalesOfferPackages = (
    salesOfferPackageInfo:
        | SelectSalesOfferPackageWithError
        | SalesOfferPackage[]
        | ProductWithSalesOfferPackages[]
        | undefined,
): salesOfferPackageInfo is SalesOfferPackage[] =>
    salesOfferPackageInfo !== undefined &&
    (salesOfferPackageInfo as SalesOfferPackage[])[0].ticketFormats !== undefined;

export const isSingleTicketProps = (
    props: SingleTicketProps | ReturnTicketProps | PeriodTicketProps | FlatFareTicketProps,
): props is SingleTicketProps => (props as SingleTicketProps).journeyDirection !== undefined;

export const isReturnTicketProps = (
    props: SingleTicketProps | ReturnTicketProps | PeriodTicketProps | FlatFareTicketProps,
): props is ReturnTicketProps => (props as ReturnTicketProps).circular !== undefined;

export const isPeriodTicketProps = (
    props: SingleTicketProps | ReturnTicketProps | PeriodTicketProps | FlatFareTicketProps,
): props is PeriodTicketProps => (props as PeriodTicketProps).zone !== undefined;

export const isFlatFareTicketProps = (
    props: SingleTicketProps | ReturnTicketProps | PeriodTicketProps | FlatFareTicketProps,
): props is FlatFareTicketProps => (props as FlatFareTicketProps).productName !== undefined;

export const isTicketPeriodAttributeWithErrors = (
    productDates: TicketPeriod | TicketPeriodWithErrors | undefined,
): productDates is TicketPeriodWithErrors =>
    productDates !== undefined && (productDates as TicketPeriodWithErrors).errors !== undefined;

export const isTicketPeriodAttribute = (
    productDates: TicketPeriod | TicketPeriodWithErrors | undefined,
): productDates is TicketPeriod =>
    productDates !== undefined &&
    (productDates as TicketPeriod).startDate !== undefined &&
    (productDates as TicketPeriod).endDate !== undefined;

export const isMultipleOperatorAttributeWithErrors = (
    searchOperator: MultipleOperatorsAttribute | MultipleOperatorsAttributeWithErrors | undefined,
): searchOperator is MultipleOperatorsAttributeWithErrors =>
    searchOperator !== undefined && (searchOperator as MultipleOperatorsAttributeWithErrors).errors !== undefined;

export const isMultiOperatorInfoWithErrors = (
    multiOperatorInfo: MultiOperatorInfo[] | MultiOperatorInfoWithErrors | undefined,
): multiOperatorInfo is MultiOperatorInfoWithErrors =>
    multiOperatorInfo !== undefined && (multiOperatorInfo as MultiOperatorInfoWithErrors).errors !== undefined;
