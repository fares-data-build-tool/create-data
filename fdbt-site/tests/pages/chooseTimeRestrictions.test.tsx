import * as React from 'react';
import { shallow } from 'enzyme';
import ChooseTimeRestrictions, { getServerSideProps } from '../../src/pages/chooseTimeRestrictions';
import { getMockContext } from '../testData/mockData';
import { FULL_TIME_RESTRICTIONS_ATTRIBUTE, TIME_RESTRICTIONS_DEFINITION_ATTRIBUTE } from '../../src/constants';
import { ErrorInfo } from '../../src/interfaces';

describe('pages', () => {
    describe('chooseTimeRestrictions', () => {
        it('should render correctly with no prior inputs', () => {
            const tree = shallow(
                <ChooseTimeRestrictions
                    chosenDays={['monday', 'tuesday']}
                    startTimeInputs={[]}
                    endTimeInputs={[]}
                    errors={[]}
                    csrfToken=""
                />,
            );
            expect(tree).toMatchSnapshot();
        });
        it('should render correctly with prior inputs ', () => {
            const tree = shallow(
                <ChooseTimeRestrictions
                    chosenDays={['monday', 'tuesday']}
                    startTimeInputs={[{ timeInput: '0900', day: 'monday' }]}
                    endTimeInputs={[{ timeInput: '1900', day: 'tuesday' }]}
                    errors={[]}
                    csrfToken=""
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render correctly with errors', () => {
            const tree = shallow(
                <ChooseTimeRestrictions
                    chosenDays={['monday', 'tuesday']}
                    startTimeInputs={[
                        { timeInput: '0000', day: 'monday' },
                        { timeInput: '0900', day: 'tuesday' },
                    ]}
                    endTimeInputs={[
                        { timeInput: '2400', day: 'monday' },
                        { timeInput: '7pm', day: 'tuesday' },
                    ]}
                    errors={[
                        {
                            errorMessage: '2400 is not a valid input. Use 0000.',
                            id: `end-time-monday`,
                            userInput: '2400',
                        },
                        {
                            errorMessage: 'Time must be in 2400 format',
                            id: `end-time-tuesday`,
                            userInput: '7pm',
                        },
                    ]}
                    csrfToken=""
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        describe('getServerSideProps', () => {
            const mockDays = ['monday', 'tuesday', 'friday'];

            it('should return default props when the page is first visited', () => {
                const expectedProps = {
                    chosenDays: mockDays,
                    errors: [],
                    csrfToken: '',
                    startTimeInputs: [],
                    endTimeInputs: [],
                };
                const ctx = getMockContext({
                    session: { [TIME_RESTRICTIONS_DEFINITION_ATTRIBUTE]: { validDays: mockDays } },
                });
                const actualProps = getServerSideProps(ctx);
                expect(actualProps.props).toEqual(expectedProps);
            });

            it('should return props with errors when the user enters incorrect time inputs', () => {
                const expectedUserInputStructure = expect.arrayContaining([
                    expect.objectContaining({ timeInput: expect.any(String), day: expect.any(String) }),
                ]);
                const mockUserInputs = [
                    { day: 'monday', startTime: '0900', endTime: 'dad' },
                    { day: 'tuesday', startTime: 'vveaee', endTime: '2300' },
                    { day: 'wednesday', startTime: '2400', endTime: '' },
                    { day: 'thursday', startTime: '', endTime: '' },
                ];
                const mockErrors: ErrorInfo[] = [
                    {
                        errorMessage: 'Time must be in 2400 format',
                        id: 'end-time-monday',
                        userInput: 'turkey',
                    },
                    {
                        errorMessage: 'Time must be in 2400 format',
                        id: 'start-time-tuesday',
                        userInput: 'vveaee',
                    },
                    {
                        errorMessage: '2400 is not a valid input. Use 0000.',
                        id: 'start-time-wednesday',
                        userInput: '2400',
                    },
                ];
                const expectedProps = {
                    chosenDays: mockDays,
                    errors: mockErrors,
                    csrfToken: '',
                    startTimeInputs: expectedUserInputStructure,
                    endTimeInputs: expectedUserInputStructure,
                };
                const ctx = getMockContext({
                    session: {
                        [TIME_RESTRICTIONS_DEFINITION_ATTRIBUTE]: { validDays: mockDays },
                        [FULL_TIME_RESTRICTIONS_ATTRIBUTE]: {
                            fullTimeRestrictions: mockUserInputs,
                            errors: mockErrors,
                        },
                    },
                });
                const actualProps = getServerSideProps(ctx);
                expect(actualProps.props).toEqual(expectedProps);
            });
        });
    });
});
