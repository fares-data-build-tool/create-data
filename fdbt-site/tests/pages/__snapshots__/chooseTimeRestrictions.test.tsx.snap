// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages chooseTimeRestrictions should render correctly with errors 1`] = `
<Component
  description="Choose time restrictions page of the Create Fares Data Service"
  errors={
    Array [
      Object {
        "errorMessage": "2400 is not a valid input. Use 0000.",
        "id": "end-time-monday",
        "userInput": "2400",
      },
      Object {
        "errorMessage": "Time must be in 2400 format",
        "id": "end-time-tuesday",
        "userInput": "7pm",
      },
    ]
  }
  title="Choose time restrictions - Create Fares Data Service "
>
  <CsrfForm
    action="/api/chooseTimeRestrictions"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={
        Array [
          Object {
            "errorMessage": "2400 is not a valid input. Use 0000.",
            "id": "end-time-monday",
            "userInput": "2400",
          },
          Object {
            "errorMessage": "Time must be in 2400 format",
            "id": "end-time-tuesday",
            "userInput": "7pm",
          },
        ]
      }
    />
    <h1
      className="govuk-heading-l"
    >
      Tell us more about the time restrictions
    </h1>
    <span
      className="govuk-hint"
    >
      Enter the times at which your ticket(s) start and end, if applicable. If they are valid at all times, leave them blank. You can leave them all blank, if needed, but you cannot enter an end time without a start time.
    </span>
    <div
      className="govuk-inset-text"
      id="time-restrictions-hint"
    >
      Enter times in 24hr format. For example 0900 is 9am, 1730 is 5:30pm.
    </div>
    <TimeRestrictionsTable
      chosenDays={
        Array [
          "monday",
          "tuesday",
        ]
      }
      dayCounters={Array []}
      endTimeInputs={
        Array [
          Object {
            "day": "monday",
            "timeInput": "2400",
          },
          Object {
            "day": "tuesday",
            "timeInput": "7pm",
          },
        ]
      }
      errors={
        Array [
          Object {
            "errorMessage": "2400 is not a valid input. Use 0000.",
            "id": "end-time-monday",
            "userInput": "2400",
          },
          Object {
            "errorMessage": "Time must be in 2400 format",
            "id": "end-time-tuesday",
            "userInput": "7pm",
          },
        ]
      }
      startTimeInputs={
        Array [
          Object {
            "day": "monday",
            "timeInput": "0000",
          },
          Object {
            "day": "tuesday",
            "timeInput": "0900",
          },
        ]
      }
    />
    <label
      className="govuk-label"
      htmlFor="time-restrictions-name"
      id="save-time-restriction-label"
    >
      <i>
        Optional - if you want to save this time restriction for reuse on other products, provide a name below
      </i>
    </label>
    <FormElementWrapper
      errorClass="govuk-input--error"
      errorId="product-details-name"
      errors={
        Array [
          Object {
            "errorMessage": "2400 is not a valid input. Use 0000.",
            "id": "end-time-monday",
            "userInput": "2400",
          },
          Object {
            "errorMessage": "Time must be in 2400 format",
            "id": "end-time-tuesday",
            "userInput": "7pm",
          },
        ]
      }
    >
      <input
        className="govuk-input govuk-input--width-30 govuk-product-name-input__inner__input"
        id="time-restrictions-name"
        maxLength={50}
        name="timeRestrictionName"
        type="text"
      />
    </FormElementWrapper>
    <input
      className="govuk-button continue-button-placement"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;

exports[`pages chooseTimeRestrictions should render correctly with no prior inputs 1`] = `
<Component
  description="Choose time restrictions page of the Create Fares Data Service"
  errors={Array []}
  title="Choose time restrictions - Create Fares Data Service "
>
  <CsrfForm
    action="/api/chooseTimeRestrictions"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <h1
      className="govuk-heading-l"
    >
      Tell us more about the time restrictions
    </h1>
    <span
      className="govuk-hint"
    >
      Enter the times at which your ticket(s) start and end, if applicable. If they are valid at all times, leave them blank. You can leave them all blank, if needed, but you cannot enter an end time without a start time.
    </span>
    <div
      className="govuk-inset-text"
      id="time-restrictions-hint"
    >
      Enter times in 24hr format. For example 0900 is 9am, 1730 is 5:30pm.
    </div>
    <TimeRestrictionsTable
      chosenDays={
        Array [
          "monday",
          "tuesday",
        ]
      }
      dayCounters={Array []}
      endTimeInputs={Array []}
      errors={Array []}
      startTimeInputs={Array []}
    />
    <label
      className="govuk-label"
      htmlFor="time-restrictions-name"
      id="save-time-restriction-label"
    >
      <i>
        Optional - if you want to save this time restriction for reuse on other products, provide a name below
      </i>
    </label>
    <FormElementWrapper
      errorClass="govuk-input--error"
      errorId="product-details-name"
      errors={Array []}
    >
      <input
        className="govuk-input govuk-input--width-30 govuk-product-name-input__inner__input"
        id="time-restrictions-name"
        maxLength={50}
        name="timeRestrictionName"
        type="text"
      />
    </FormElementWrapper>
    <input
      className="govuk-button continue-button-placement"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;

exports[`pages chooseTimeRestrictions should render correctly with prior inputs  1`] = `
<Component
  description="Choose time restrictions page of the Create Fares Data Service"
  errors={Array []}
  title="Choose time restrictions - Create Fares Data Service "
>
  <CsrfForm
    action="/api/chooseTimeRestrictions"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <h1
      className="govuk-heading-l"
    >
      Tell us more about the time restrictions
    </h1>
    <span
      className="govuk-hint"
    >
      Enter the times at which your ticket(s) start and end, if applicable. If they are valid at all times, leave them blank. You can leave them all blank, if needed, but you cannot enter an end time without a start time.
    </span>
    <div
      className="govuk-inset-text"
      id="time-restrictions-hint"
    >
      Enter times in 24hr format. For example 0900 is 9am, 1730 is 5:30pm.
    </div>
    <TimeRestrictionsTable
      chosenDays={
        Array [
          "monday",
          "tuesday",
        ]
      }
      dayCounters={Array []}
      endTimeInputs={
        Array [
          Object {
            "day": "tuesday",
            "timeInput": "1900",
          },
        ]
      }
      errors={Array []}
      startTimeInputs={
        Array [
          Object {
            "day": "monday",
            "timeInput": "0900",
          },
        ]
      }
    />
    <label
      className="govuk-label"
      htmlFor="time-restrictions-name"
      id="save-time-restriction-label"
    >
      <i>
        Optional - if you want to save this time restriction for reuse on other products, provide a name below
      </i>
    </label>
    <FormElementWrapper
      errorClass="govuk-input--error"
      errorId="product-details-name"
      errors={Array []}
    >
      <input
        className="govuk-input govuk-input--width-30 govuk-product-name-input__inner__input"
        id="time-restrictions-name"
        maxLength={50}
        name="timeRestrictionName"
        type="text"
      />
    </FormElementWrapper>
    <input
      className="govuk-button continue-button-placement"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;
